packages/remirror__extension-file/src/file-upload-context.ts

```diff
@@ -0,0 +1,43 @@
+export interface UploadContext {
+  set: (key: string, value: unknown) => void;
+  get: (key: string) => unknown;
+  addListener: (listener: UploadContextListener) => void;
+}
+
+type UploadContextListener = (values: Record<string, unknown>) => void;
+
+export function createUploadContext(): UploadContext {
+  const values: Record<string, unknown> = {};
+  const eventListeners: UploadContextListener[] = [];
+  let timeoutId: ReturnType<typeof setTimeout> | null = null;
+
+  const get = (key: string): unknown => {
+    return values[key];
+  };
+
+  const set = (key: string, value: unknown) => {
+    values[key] = value;
+    afterSet();
+  };
+
+  const addListener = (listener: UploadContextListener) => {
+    eventListeners.push(listener);
+  };
+
+  const afterSet = () => {
+    if (timeoutId) {
+      clearTimeout(timeoutId);
+    }
+
+    timeoutId = setTimeout(() => {
+      eventListeners.forEach((listener) => listener(values));
+      timeoutId = null;
+    });
+  };
+
+  return {
+    set,
+    get,
+    addListener: addListener,
```

Comment:
Do we add a `removeListener` here too?

Reply:
`addListener` will returns a callback that will remove the listener. Thanks to `nanoevents` this is so easy.